lib/miq_expression.rb

```diff
@@ -598,45 +598,47 @@ def self.value2human(val, options = {})
     ret
   end
 
-  def self.operands2rubyvalue(operator, ops, context_type)
-    # puts "Enter: operands2rubyvalue: operator: #{operator}, ops: #{ops.inspect}"
+  UNQUOTABLE_OPERATORS = %w[= IS\ NULL IS\ NOT\ NULL IS\ EMPTY IS\ NOT\ EMPTY].freeze
 
-    if ops["field"]
-      if ops["field"] == "<count>"
-        ["<count>", quote(ops["value"], "integer")]
-      else
-        target = parse_field_or_tag(ops["field"])
-        col_type = target&.column_type || "string"
-        if context_type == "hash"
-          val = ops["field"].split(".").last.split("-").join(".")
-          fld = "<value type=#{col_type}>#{val}</value>"
-        else
-          fld = "<value ref=#{target.model.to_s.downcase}, type=#{col_type}>#{target.tag_path_with}</value>"
-        end
-        if ["like", "not like", "starts with", "ends with", "includes", "regular expression matches", "regular expression does not match"].include?(operator)
-          [fld, ops["value"]]
+  def self.quote_by(operator, value, column_type = nil)
+    if (STRING_OPERATORS + DEPRECATED_OPERATORS - UNQUOTABLE_OPERATORS).map(&:downcase).include?(operator)
+      value
+    else
+      quote(value, column_type.to_s)
+    end
+  end
+
+  def self.operands2rubyvalue(operator, ops, context_type)
+    field, operator, value, column_type =
+      if ops["field"]
+        if ops["field"] == "<count>"
+          ["<count>", nil, ops["value"], :integer]
         else
-          [fld, quote(ops["value"], col_type.to_s)]
+          target = parse_field_or_tag(ops["field"])
+          col_type = target&.column_type || "string"
+          field = if context_type == "hash"
+                    val = ops["field"].split(".").last.split("-").join(".")
+                    "<value type=#{col_type}>#{val}</value>"
+                  else
+                    "<value ref=#{target.model.to_s.downcase}, type=#{col_type}>#{target.tag_path_with}</value>"
+                  end
+
+          [field, operator, ops["value"], col_type]
```

Comment:
I think calling `quote_by(operator, ops["value"], col_type)` first and then having `if context_type == "hash"` just return the values (not setting `field`) will cut down on the complexity here.

Reply:
I did some update here but I am not sure whether it is what you suggested